### PR TITLE
Refactor function type logic

### PR DESCRIPTION
**Pull request by nicolasstucki**
* Rename `isFucntionType` to `isFunctionNType` 
  (tests for FunctionN/ContextFunctionN)
* Rename `isFunctionOrPolyType` to `isFunctionType` 
  (tests for FunctionN/ContextFunctionN/PolyFunction/ErasedFunction)
* Rename `isRefinedFunctionType` to `isPolyOrErasedFunctionType`
  (tests for PolyFunction/ErasedFunction)
* Define `isPolyFunctionType`
* Refactor uses of `isPolyFunctionType`/`isErasedFunctionType`/`isPolyOrErasedFunctionType`
* Simplify/generalize erasure of refined function types

Followup of https://github.com/dotty-staging/dotty/commit/6a349b50f90db3f9a279e0399e6ab44ff11ed496#r121364261

### CODE REVIEW

**Comment by nicolasstucki** [comment]
We should double-check all uses of `isFunctionNType` to determine which ones might actually need `isFunctionType`.

**Review by smarter — APPROVED** [review]
Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -1888,6 +1888,9 @@ object Types {
         if alwaysDependent || mt.isResultDependent then
           RefinedType(funType, nme.apply, mt)
         else funType
+      case mt: MethodType if !mt.isParamDependent =>
+        assert(mt.hasErasedParams)
+        RefinedType(defn.ErasedFunctionType, nme.apply, mt)
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: I also moved `!mt.isParamDependent` into assertions to make it clear that this is unexpected.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 2]
```diff
@@ -666,15 +666,17 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
-                case info1: PolyType =>
-                  return isSubInfo(info1, tp2.refinedInfo)
-                case _ =>
+              if defn.isPolyFunctionType(tp2) then
+                // TODO should we handle ErasedFunction is this same way?
+                tp1.member(nme.apply).info match
+                  case info1: PolyType =>
+                    return isSubInfo(info1, tp2.refinedInfo)
+                  case _ =>
+              else
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:675]: I am not sure why there is a difference between the two comparisons. Would it work to always use the second one, which tests for `RefinedType`?

**Review by odersky — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:675]: I will try to merge the two cases into one.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 3]
```diff
@@ -1709,21 +1710,29 @@ class Definitions {
    *  - scala.FunctionN
    *  - scala.ContextFunctionN
    */
-  def isFunctionType(tp: Type)(using Context): Boolean =
+  def isFunctionNType(tp: Type)(using Context): Boolean =
     isNonRefinedFunction(tp.dropDependentRefinement)
 
   /** Is `tp` a specialized, refined function type? Either an `ErasedFunction` or a `PolyFunction`. */
   def isRefinedFunctionType(tp: Type)(using Context): Boolean =
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala]: Used `isPolyOrErasedFunctionType`

**Pull request merged**